### ISSUE TITLE
Why does Publish call OnPublish ?

### ISSUE THREAD

**Issue by Libong** [mochi-co/mqtt#249]
What is the meaning of this code? If I call the Publish or InjectPacket function, will the content sent by the server be passed back to OnPublish for the server to receive

if pkx, err := s.hooks.OnPublish(cl, pk); err == nil {
		pk = pkx
	} else if errors.Is(err, packets.ErrRejectPacket) {
		return nil
	}

**Comment by thedevop** [comment]
Yes, OnPublish will be called. You can filter out in the hook by checking for `cl.Net.Inline`.

**Comment by Libong** [comment]
I see in the code that only when the Publish function that encapsulates InjectPacket is issued, will the inline be set to true. If I call InjectPacket instead of Publish, how can I filter it? Alternatively, you can only call the Publish function

**Comment by thedevop** [comment]
Publish is merely a convenient function that wrap InjectPacket.
https://github.com/mochi-co/mqtt/blob/0648e395075ee3072747d259e777b7166619217c/server.go#L647-L649

If you directly use InjectPacket, you should also use an inline client.

To filter out inline publish in hook:
```go
	if cl.Net.Inline {
		return pk, nil
	}
```

**Comment by oherych** [comment]
Hi. I have additional questions about this code. Why do we ignore errors? What should be in case if we have have error different from `packets.ErrRejectPacket`. For example in my case I want to send `packets.ErrPayloadFormatInvalid`

```go
	if pkx, err := s.hooks.OnPublish(cl, pk); err == nil {
		pk = pkx
	} else if errors.Is(err, packets.ErrRejectPacket) {
		return nil
	}
```

**Comment by mochi-co** [comment]
@oherych `packets.ErrRejectPacket` is used in order to silently drop/reject an incoming packet. 

The [MQTTv5 Spec](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901120) doesn't detail a standardised mechanism for returning error messages to clients. That said - QoS > 0 packets do return Puback/Pubrec packets on receipt of a QoS packet, and in MQTTv5 both the reason code and reason string may be set in response.

In your case, you could use `cl.WritePacket` to issue a an ack with the packets.ErrPayloadFormatInvalid error code and reason string the client, and then return a `packets.ErrRejectPacket` as the err response for `s.hooks.OnPublish`. For example, something like this:

```go
packets.Packet{
	FixedHeader: packets.FixedHeader{
        Type: packets.Puback,
        Qos:  1,
    },
    PacketID:   pk.PacketID,
    ReasonCode: packets.ErrPayloadFormatInvalid.Code, 
    Properties: packets.Properties{
        ReasonString: packets.ErrPayloadFormatInvalid.Reason,
    },
}
```

This is a bit of a roundabout way of achieving it. I recognise that with the implementation of hooks, we should be able to return an error ack in response to a publish message (if that publish is QoS>0). As such I have created a PR to add this capability, and fix a related bug. If it passes review, it will likely go out in the weekly release (typically weekends when I have some time).

**Comment by mochi-co** [comment]
@oherych This has been resolved in [v2.2.16](https://github.com/mochi-co/mqtt/releases/tag/v2.2.16), plus the addition of the new `packets.CodeSuccessIgnore` packet - you should be able to now return any `packets.Code` error from OnPublish and get an error ack.
